internal/mesh/internal/controllers/sidecarproxy/controller_test.go

```diff
@@ -859,104 +859,112 @@ func TestMeshController(t *testing.T) {
-					break
+	suite.waitForProxyStateTemplateState(t, tmplResource.Id, func(t resourcetest.T, tmpl *pbmesh.ProxyStateTemplate) {
+		// Check outbound listener.
+		var foundListener bool
+		for _, l := range tmpl.ProxyState.Listeners {
+			if strings.Contains(l.Name, xdscommon.OutboundListenerName) && l.Direction == pbproxystate.Direction_DIRECTION_OUTBOUND {
+				foundListener = true
+
+				// Check the listener filter chain
+				for _, r := range l.Routers {
+					foundByName := false
+					switch x := r.Destination.(type) {
+					case *pbproxystate.Router_L4:
+						// TcpProxy is easy, so just having this exist is
+						// sufficient. We don't have to deep inspect it. We care
+						// that there IS a listener matching the destination. If
+						// there is a TCPRoute with a split or a rename we don't
+						// care here.
+						foundByName = true
+					case *pbproxystate.Router_L7:
+						require.NotNil(t, x.L7.Route)
+						routerName := x.L7.Route.Name
+						foundByName = strings.Contains(routerName, name)
+					default:
+						t.Fatalf("unexpected type of destination: %T", r.Destination)
+					}
+
+					if foundByName {
+						// We expect that there is a filter chain match for transparent proxy destinations.
+						require.NotNil(t, r.Match)
+						require.NotEmpty(t, r.Match.PrefixRanges)
+						break
+					}
 				}
+				break
 			}
-			break
 		}
-	}
-	require.True(t, foundListener)
+		require.True(t, foundListener)
+	})
 
-	requireClustersAndEndpoints(t, name, &tmpl, true)
+	suite.requireClustersAndEndpoints(t, name, tmplResource, true)
 }
 
-func requireClustersAndEndpoints(t *testing.T, name string, tmpl *pbmesh.ProxyStateTemplate, found bool) {
+func (suite *controllerTestSuite) requireClustersAndEndpoints(t *testing.T, name string, tmplResource *pbresource.Resource, found bool) {
 	t.Helper()
 
-	var foundCluster bool
-	for c := range tmpl.ProxyState.Clusters {
-		if strings.Contains(c, name) {
-			foundCluster = true
-			break
+	suite.waitForProxyStateTemplateState(t, tmplResource.Id, func(t resourcetest.T, tmpl *pbmesh.ProxyStateTemplate) {
+		var foundCluster bool
+		for c := range tmpl.ProxyState.Clusters {
+			if strings.Contains(c, name) {
+				foundCluster = true
+				break
+			}
 		}
-	}
 
-	require.Equal(t, found, foundCluster)
+		require.Equal(t, found, foundCluster)
 
-	var foundEndpoints bool
-	for c := range tmpl.RequiredEndpoints {
-		if strings.Contains(c, name) {
-			foundEndpoints = true
-			break
+		var foundEndpoints bool
+		for c := range tmpl.RequiredEndpoints {
+			if strings.Contains(c, name) {
+				foundEndpoints = true
+				break
+			}
 		}
-	}
 
-	require.Equal(t, found, foundEndpoints)
+		require.Equal(t, found, foundEndpoints)
+	})
+}
+
+func (suite *controllerTestSuite) waitForProxyStateTemplateState(t *testing.T, id *pbresource.ID, verify func(t resourcetest.T, tmpl *pbmesh.ProxyStateTemplate)) {
+	suite.client.WaitForResourceState(t, id, func(t resourcetest.T, res *pbresource.Resource) {
```

Comment:
Could we rename some of the `t`'s floating around to something like `tt` or `rt`? It's hard to tell at a glance what's being shadowed

Reply:
Sure, good call 👍🏻